seqr/views/apis/dashboard_api_tests.py

```diff
@@ -120,3 +126,57 @@ def test_export_projects_table(self):
         response = self.client.get('{}?file_format=csv'.format(url))
         self.assertEqual(response.status_code, 400)
         self.assertDictEqual(response.json(), {'error': 'Invalid file_format: csv'})
+
+
+if anvil_enabled():
+    @mock.patch('seqr.views.utils.terra_api_utils.GOOGLE_SERVICE_ACCOUNT_INFO', GOOGLE_SERVICE_ACCOUNT_INFO)
+    class DashboardPageAnvilTest(DashboardPageTest):
```

Comment:
So this code style is a bit hard to read, and I think it will be really difficult to extend this without tons of copy/paste to other places that also need to have full testing for the Anvil case  (i.e. the api calls to get all the collaborators for a project). What I'm proposing below is one way to update the tests to effectively run tests against both login modes, but I  am happy to meet to discuss if you don't like this approach or have any questions.

I think this could work a lot better/ be a lot more extensible to other test files if instead of linear inheritance of `DashboardPageAnvilTest` -> `DashboardPageTest` -> `AuthenticationTestCase` you instead used a mixin style. So what that would look like is you define an `AnvilAuthenticationTestCase` that inherits from `AuthenticationTestCase` but completely redefines defines the `_check_login` method and the `setUpTestData` to instead work with AnVIL login and not add them to user groups for permissions. This class will be where you mock out the `TERRA_API_ROOT_URL ` to make sure it is defined. You would ideally set up the users to have the same permissions in both (i.e. making sure that `collaborator_user` has the same project collaborator permissions in the mocked anvil as the mocked local groups).  Then instead of defining the main class as `class DashboardPageTest(AuthenticationTestCase)` you would do
```
class DashboardPageTest:
   <basically same code as now>
   ...

class LocalUserDashboardPageTest(AuthenticationTestCase, DashboardPageTest)

class AnvilDashboardPageTest(AnvilAuthenticationTestCase, DashboardPageTest):
   def test_dashboard_page_data(self):
        super(self, AnvilDashboardPageTest).test_dashboard_page_data()
       <anvil specific tests>
```

If you are struggling to get `responses` to work in the set up for the `AnvilAuthenticationTestCase`, you could instead patch the helper methods that make calls to AnVIL, and then just make sure that those functions are really well tested. So for example, in the `AnvilAuthenticationTestCase` define 
```
USER_WORKSPACE_MAP = {<a mapping between test users and the response bodies for what workspaces we want them to have access to>}
...
def setUp():
   patcher = mock.patch('seqr.views.utils.terra_api_utils.list_anvil_workspaces')
   patcher.start().side_effect = lambda user, *args: USER_WORKSPACE_MAP[user.email]
   self.addCleanup(patcher.stop)
```

Reply:
Very heuristic idea. I have never think about doing it in this way. I still like to meet you next week for some questions.

Reply:
Some thoughts for the meeting this afternoon:  
We need three kinds of tests with different permission settings: `pure local permissions`, `combined local and remote permissions`, and `pure remote permissions`. Some highlights of these tests are listed below.  
  
- Pure local permissions  
   - Test without `TERRA_API_ROOT_URL`  
   - All existing test cases should work  
  
- Combine local and remote permissions  
   - Test with `TERRA_API_ROOT_URL`  
   -  All existing test cases should work
   - Add a few new test cases for some corners to make sure:   
      - the users still have the existing permissions  
      - new permissions are added.  
  
- Pure remote permissions.  
   - Test with `TERRA_API_ROOT_URL`   
   - Provide a new `AnvilAuthenticationTestCase` class to:   
      - set all the local permissions to the mocked remote permissions  
   - Ideally, re-use all existing test cases

Reply:
This is a really good summary of what we need. Looking forward to discussing